conformance/lang/expressions/let-expr/let_expr.balt

```diff
@@ -0,0 +1,431 @@
+    string city;
+    string? country;
+};
+
+function init() {
+    io:println(let table<map<any>> tbl = table [{ id: 13 , name: "Emm", lname: "Dean" }] in tbl);
+    // @output [{"id":13,"name":"Emm","lname":"Dean"}]
+    io:println(() == let table<Foo> key(m) tbl = table [{m: {"AAA": "DDDD"}, age: 31 }]
+        in tbl.add({ m: {"CCC":"EEEE"}, age: 34 })); // @output true
+    io:println(let table<Student> key(name) tbl = table [{name: "Amy", id: 1234, address:
+        {"street": "Main Street", "city": "Colombo", "country": ()}}] in tbl["Amy"]["address"]["country"] == ());
+    // @output true
+}
+
+Test-Case: output
+Description: Test let-expr with query expressions.
+Labels: let-expr, record-type, array-type, list-constructor-expr
+
+type Department record {|
+   int id;
+   string name;
+|};
+
+type Person record {|
+   int id;
+   string fname;
+   string lname;
+|};
+
+type DeptPerson record {|
+   string fname;
+   string lname;
+   string? dept;
+|};
+
+Person[] personList = [
+    {id: 1, fname: "Alex", lname: "George"}
+];
+
+Department[] deptList = [
+    {id: 1, name:"HR"}
+];
+
+function init() {
+    var l = let Person[] personList = [{id: 1, fname: "Sam", lname: "George"}],
+                   Department[] deptList = [{id: 1, name:"HR"}] in
+                   from var person in personList
+                        join var {id: deptId, name: deptName} in deptList
+                        on person.id equals deptId
+                        select {
+                            fname : person.fname,
+                            lname : person.lname,
+                            dept : deptName
+                        };
+    io:println(l); // @output [{"fname":"Sam","lname":"George","dept":"HR"}]
+}
+
+Test-Case: parser-error
+Description: let-expr with missing tokens
+Labels: let-expr
+
+function errorFunction() {
+    io:println(let int i in i); // @error missing equal token
+                                // @error missing identifier
+    io:println(let i = 2 in i); // @error unknown type 'i'
+                                // @error missing identifier
+                                // @error undefined symbol 'i'
+    io:println(let int i = 2 i); // @error missing in keyword
+    io:println(let int i = 2 int j = 2 in i + j); // @error missing comma token
+    io:println(let in i); // @error missing let variable declaration
+                          // @error undefined symbol 'i'
+    io:println(let i); // @error unknown type 'i'
+                       // @error missing equal token
+                       // @error missing identifier
+                       // @error missing identifier
+                       // @error missing identifier
+                       // @error missing in keyword
+}
+
+Test-Case: error
+Description: let-expr when variables are declared outside its' scope
+Labels: let-expr
+
+function errorFunction() {
+    int i = 2;
+    io:println(let int i = 2 in i); // @error redeclared symbol 'i'
+}
```

Comment:
Shall we add some `error` tests to verify the following section of the spec?

> The typed-binding-pattern is used unconditionally, meaning that it is a compile error if the static types do not guarantee the success of the match. If the typed-binding-pattern uses var, then the type of the variable is inferred from the precise static type of the expression following =.

Reply:
Added